src/main/java/build/buildfarm/common/resources/ResourceParser.java

```diff
@@ -0,0 +1,201 @@
+
+    // Extract each of the following segments:
+    // `{instance_name}/uploads/{uuid}/blobs/{hash}/{size}{/optional_metadata}` or
+    // `{instance_name}/uploads/{uuid}/compressed-blobs/{compressor}/{uncompressed_hash}/{uncompressed_size}{/optional_metadata}`
+    UploadBlobRequest.Builder builder = UploadBlobRequest.newBuilder();
+    builder.setInstanceName(
+        String.join("/", Arrays.copyOfRange(segments, 0, index.getAndIncrement())));
+    builder.setUuid(segments[index.getAndIncrement()]);
+    builder.setBlob(parseBlobInformation(segments, index));
+    if (index.intValue() < segments.length) {
+      builder.setMetadata(
+          String.join("/", Arrays.copyOfRange(segments, index.intValue(), segments.length)));
+    }
+    return builder.build();
+  }
+  /**
+   * @brief Parse the resource name into a specific type.
+   * @details Assumes the URI is already is already in the download blob format.
+   * @param resourceName The resource name URI defined by REAPI.
+   * @return The resource type derived from the URI.
+   * @note Suggested return identifier: request.
+   */
+  public static DownloadBlobRequest parseDownloadBlobRequest(String resourceName) {
+    // Find the index of the keyword.  This will give us an initial index to extract information
+    // from.
+    String[] segments = tokenize(resourceName);
+    MutableInt index = startIndex(segments, Resource.TypeCase.DOWNLOAD_BLOB_REQUEST);
+
+    // Extract each of the following segments: `{instance_name}/blobs/{hash}/{size}` or
+    // `{instance_name}/compressed-blobs/{compressor}/{uncompressed_hash}/{uncompressed_size}`
+    DownloadBlobRequest.Builder builder = DownloadBlobRequest.newBuilder();
+    builder.setInstanceName(String.join("/", Arrays.copyOfRange(segments, 0, index.intValue())));
+    builder.setBlob(parseBlobInformation(segments, index));
+    return builder.build();
+  }
+  /**
+   * @brief A mapping of REAPI keywords to their respective resource types.
+   * @details This lookup table can be used for parsing.
+   * @return A lookup table for keyword to resource type.
+   * @note Suggested return identifier: keywordMap.
+   */
+  private static HashMap<String, Resource.TypeCase> keywordResourceMap() {
+    HashMap<String, Resource.TypeCase> keywords = new HashMap<>();
+    keywords.put("blobs", Resource.TypeCase.DOWNLOAD_BLOB_REQUEST);
+    keywords.put("compressed-blobs", Resource.TypeCase.DOWNLOAD_BLOB_REQUEST);
+    keywords.put("uploads", Resource.TypeCase.UPLOAD_BLOB_REQUEST);
+    keywords.put("operations", Resource.TypeCase.STREAM_OPERATION_REQUEST);
+    return keywords;
+  }
+  /**
+   * @brief Tokenize the resource name URI.
+   * @details A simple split should be fine.
+   * @param resourceName The resource name URI defined by REAPI.
+   * @return A tokenized resource name.
+   * @note Suggested return identifier: uRITokens.
+   */
+  private static String[] tokenize(String resourceName) {
+    return resourceName.split("/");
+  }
```

Comment:
good place to put the neighbor `resourcePath(Iterable)` method

Reply:
I'm thinking what the implementation would be.  I wasn't able to parse the resource path sequentially.

Reply:
Just the ever present join analogue to `tokenize`

```
private static String resourcePath(Iterable components) {
  return RESOURCE_SEPARATOR.join(components);
}
```

Reply:
Replaced array argument with `Iterable<String>`.  Removed original implementation since it wouldn't get test coverage.